bpf/lib/nodeport.h

```diff
@@ -695,6 +696,17 @@ int tail_nodeport_ipv6_dsr(struct __ctx_buff *ctx)
 #elif DSR_ENCAP_MODE == DSR_ENCAP_NONE
 	ret = dsr_set_ext6(ctx, ip6, &addr, port, &ohead);
 #elif DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
+
+	ret = maybe_add_l2_hdr(ctx, ENCAP_IFINDEX, &l2_hdr_required);
+	if (ret != 0)
+		goto drop_err;
+
+	if (l2_hdr_required && ETH_HLEN == 0)
+		if (!revalidate_data_l3_off(ctx, &data, &data_end, &ip6, __ETH_HLEN)) {
+			ret = DROP_INVALID;
+			goto drop_err;
+		}
+
```

Comment:
drive-by - I think we can simplify this a bit more and avoid the dependency on ENCAP_IFINDEX.

The helper is generic enough to check for "are we redirecting from L3 to L3?", and so requires the target interface as parameter. But we *know* that the target interface needs a L2 header (or more correctly, the desired VXLAN / GENEVE encapsulation format uses inner L2 headers). So all that's left to check is whether we are currently processing the packet on a L3 device (and thus need to add a L2 header before encapsulating).

This would also match the XDP case more closely - here we *don't* redirect to another interface. So whatever target interface we're passing in, it doesn't actually matter.